components/datetime/src/timezone.rs

```diff
@@ -0,0 +1,331 @@
+
+        for resource_key in required_time_zone_keys(&pattern)? {
+            if resource_key.eq(&provider::key::TIMEZONE_EXEMPLAR_CITIES_V1) {
+                load_resource(&locale, resource_key, &mut exemplar_cities, zone_provider)?;
+            } else if resource_key.eq(&provider::key::TIMEZONE_GENERIC_NAMES_LONG_V1) {
+                load_resource(&locale, resource_key, &mut mz_generic_long, zone_provider)?;
+            } else if resource_key.eq(&provider::key::TIMEZONE_GENERIC_NAMES_SHORT_V1) {
+                load_resource(&locale, resource_key, &mut mz_generic_short, zone_provider)?;
+            } else if resource_key.eq(&provider::key::TIMEZONE_SPECIFIC_NAMES_SHORT_V1) {
+                load_resource(&locale, resource_key, &mut mz_specific_short, zone_provider)?;
+            } else if resource_key.eq(&provider::key::TIMEZONE_SPECIFIC_NAMES_LONG_V1) {
+                load_resource(&locale, resource_key, &mut mz_specific_long, zone_provider)?;
+            }
+        }
+
+        Ok(Self {
+            pattern,
+            zone_formats,
+            exemplar_cities,
+            mz_generic_long,
+            mz_generic_short,
+            mz_specific_long,
+            mz_specific_short,
+        })
+    }
+
+    pub fn format<'s: 'd, T>(&'s self, value: &'s T) -> FormattedTimeZone<'s, T>
+    where
+        T: TimeZoneInput,
+    {
+        FormattedTimeZone {
+            time_zone_format: self,
+            time_zone: value,
+        }
+    }
+
+    pub fn format_to_write(
+        &self,
+        w: &mut impl std::fmt::Write,
+        value: &impl TimeZoneInput,
+    ) -> std::fmt::Result {
+        timezone::write_pattern(self, value, w).map_err(|_| std::fmt::Error)
+    }
+
+    pub fn format_to_string(&self, value: &impl TimeZoneInput) -> String {
+        let mut s = String::new();
+        self.format_to_write(&mut s, value)
+            .expect("Failed to write to a String.");
+        s
+    }
+
+    /// Returns the time zone in generic location format as defined by the UTS-35 spec.
+    /// e.g. France Time
+    /// https://unicode.org/reports/tr35/tr35-dates.html#Time_Zone_Format_Terminology
+    pub(super) fn generic_location_format(
+        &self,
+        time_zone: &impl TimeZoneInput,
+    ) -> Option<Cow<str>> {
+        self.exemplar_city(time_zone)
+            .map(|location| Cow::Owned(self.zone_formats.region_format.replace("{0}", &location)))
+    }
+
+    /// Returns the time zone in short generic non-location format as defined by the UTS-35 spec.
+    /// e.g. PT
+    /// https://unicode.org/reports/tr35/tr35-dates.html#Time_Zone_Format_Terminology
+    pub(super) fn short_generic_non_location_format(
+        &self,
+        time_zone: &impl TimeZoneInput,
+    ) -> Option<Cow<str>> {
+        self.mz_generic_short
+            .as_ref()
+            .and_then(|metazones| time_zone.metazone_id().and_then(|mz| metazones.get(mz)))
+            .map(Clone::clone)
```

Comment:
Thought: this is calling Clone on the Cow, which doesn't actually do anything if the data is borrowed.

Reply:
This was needed since `metazones.get(mz)` returns a `&Cow<str>`.

Without the clone, that makes the return type of this function an `Option<&Cow<str>>`, which would be fine if not for `localized_gmt_format()` which just returns a `Cow<str>` that it builds within the function, so it can't return a reference from its own stack frame.

So the clone is necessary to go from `Otpion<&Cow<str>>` to `Option<Cow<str>>`, and the fact that it "doesn't actually do anything if the data is borrowed" was intended.

For now, I've refactored the `get()` function on all of the V1 structs to just clone the borrowed Cow there, instead of returning a reference. It's the same thing, but fewer lines of code. 

If you've got a more clever solution, I'd be happy to entertain it.

Reply:
My "more clever solution" involves waiting for Zibi's streaming pattern interpolator.  Nothing much to do right now.